src/connector/src/sink/mqtt.rs

```diff
@@ -239,24 +352,256 @@ impl Drop for MqttSinkWriter {
 struct MqttSinkPayloadWriter {
     // connection to mqtt, one per executor
     client: rumqttc::v5::AsyncClient,
-    config: MqttConfig,
+    topic: Option<String>,
     qos: QoS,
     retain: bool,
+    topic_index_path: Vec<usize>,
 }
 
-impl FormattedSink for MqttSinkPayloadWriter {
-    type K = Vec<u8>;
-    type V = Vec<u8>;
+impl MqttSinkPayloadWriter {}
+
+impl MqttSinkPayloadWriter {
+    async fn write_chunk(&mut self, chunk: StreamChunk, encoder: &RowEncoderWrapper) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            if op != Op::Insert {
+                continue;
+            }
+
+            let topic = match get_topic_from_index_path(
+                &self.topic_index_path,
+                self.topic.as_deref(),
+                &row,
+            ) {
+                Some(s) => s,
+                None => {
+                    tracing::error!("topic field not found in row, skipping: {:?}", row);
+                    return Ok(());
+                }
+            };
+
+            let v = encoder.encode(row)?;
 
-    async fn write_one(&mut self, _k: Option<Self::K>, v: Option<Self::V>) -> Result<()> {
-        match v {
-            Some(v) => self
-                .client
-                .publish(&self.config.common.topic, self.qos, self.retain, v)
+            self.client
+                .publish(topic, self.qos, self.retain, v)
                 .await
                 .context("mqtt sink error")
-                .map_err(SinkError::Mqtt),
-            None => Ok(()),
+                .map_err(SinkError::Mqtt)?;
         }
+
+        Ok(())
+    }
+}
+
+fn get_topic_from_index_path<'s>(
+    path: &[usize],
+    default_topic: Option<&'s str>,
+    row: &'s RowRef<'s>,
+) -> Option<&'s str> {
+    if let Some(topic) = default_topic
+        && path.is_empty()
+    {
+        Some(topic)
+    } else {
+        let mut iter = path.iter();
+        let scalar = iter
+            .next()
+            .and_then(|pos| row.datum_at(*pos))
+            .and_then(|d| {
+                iter.try_fold(d, |d, pos| match d {
+                    ScalarRefImpl::Struct(struct_ref) => {
+                        struct_ref.iter_fields_ref().nth(*pos).flatten()
+                    }
+                    _ => None,
+                })
+            });
+        match scalar {
+            Some(ScalarRefImpl::Utf8(s)) => Some(s),
+            _ => {
+                if let Some(topic) = default_topic {
+                    Some(topic)
+                } else {
+                    None
+                }
+            }
+        }
+    }
+}
+
+fn get_topic_field_index_path<'s>(
```

Comment:
How about returning `Result<Vec<usize>>`?

In this method we can check whether the internal intermediate is struct type and the leaf data type is varchar and return with more detailed error message, and then we don't need to return and check the datatype after returning.

Reply:
Makes sense, I'm off this week but will make the changes once I'm back